src/extension/serializer.ts

```diff
@@ -844,43 +854,302 @@ export class GrpcSerializer extends SerializerBase {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    token: CancellationToken,
+  ): Promise<Uint8Array> {
+    const marshalFrontmatter = this.lifecycleIdentity === RunmeIdentity.ALL
+
+    const notebookProto = this.notebookDataToProto(data)
+    const notebook = Marshal.notebook(data, notebookProto, {
+      marshalFrontmatter,
+    }) as typeof this.protoNotebookType
+
+    // cast so when we serialize we don't get missing methods due to it not being a real proto
+    notebook.frontmatter = new es_proto.Frontmatter(notebook.frontmatter)
+
+    if (marshalFrontmatter) {
+      data.metadata ??= {}
+      data.metadata[RUNME_FRONTMATTER_PARSED] = notebook.frontmatter
+    }
+
+    const cacheId = SerializerBase.getDocumentCacheId(data.metadata)
+    this.notebookDataCache.set(cacheId as string, data)
+
+    const serialRequest = this.getSerializationRequest({ notebook })
+
+    const maskedNotebookProto = this.notebookDataToProto(data)
+    const maskedNotebook = Marshal.notebook(data, maskedNotebookProto, {
+      marshalFrontmatter,
+    }) as typeof this.protoNotebookType
+    maskedNotebook.frontmatter = new es_proto.Frontmatter(maskedNotebook.frontmatter)
+    const cacheOutputs = this.cacheNotebookOutputs(notebook, maskedNotebook, cacheId)
+    const request = this.client!.serialize(serialRequest)
+
+    // run in parallel
+    const [response] = await Promise.all([request, cacheOutputs])
+
+    if (cacheId) {
+      await this.saveNotebookOutputsByCacheId(cacheId)
+    }
+
+    if (response.result === undefined) {
+      throw new Error('serialization of notebook failed')
+    }
+
+    return response.result
+  }
+
+  // await (ser as any).cacheNotebookOutputs(fixture, 'irrelevant') ???
+  private async cacheNotebookOutputs(
+    notebook: typeof this.protoNotebookType,
+    maskedNotebook: typeof this.protoNotebookType,
+    cacheId: string | undefined,
+  ): Promise<void> {
+    let session: es_proto.RunmeSession | undefined
+    const docUri = this.cacheDocUriMapping.get(cacheId ?? '')
+    const sid = this.kernel.getRunnerEnvironment()?.getSessionId()
+    if (sid && docUri) {
+      const relativePath = path.basename(docUri.fsPath)
+      session = {
+        id: sid,
+        document: { relativePath },
+      } as es_proto.RunmeSession
+    }
+
+    const outputs = { enabled: true, summary: true }
+    const options = {
+      outputs,
+      session,
+    } as es_proto.SerializeRequestOptions
+
+    maskedNotebook.cells.forEach((cell) => {
+      cell.value = maskString(cell.value)
+      cell.outputs.forEach((out) => {
+        out.items.forEach((item) => {
+          if (item.mime === OutputType.stdout) {
+            const outDecoded = Buffer.from(item.data).toString('utf8')
+            item.data = Buffer.from(maskString(outDecoded))
+          }
+        })
+      })
+    })
+
+    const plainReq = this.getSerializationRequest({ notebook, options })
+    const plainRes = this.client!.serialize(plainReq)
```

Comment:
Do you need to make two requests one for the masked data and one for the unmasked requests?
What does the original code do?
https://github.com/stateful/vscode-runme/blob/c3c135e1c47351f0e56f266efde296b354d26f5e/src/extension/serializer.ts#L756
Is it only making a single request?

Reply:
the original code makes two requests.  I assume this is because the masked version has had parts modified and thus is not the same notebook.